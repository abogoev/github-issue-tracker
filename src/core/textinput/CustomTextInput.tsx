import React, { FC, useRef, useState } from "react";
import {
  StyleSheet,
  TextInput,
  View,
  Text,
  ViewProps,
  TextInputProps,
} from "react-native";
import theme from "../../theme/theme";

interface Props extends ViewProps {
  label: string;
  textInputProps?: TextInputProps;
  error?: string | boolean;
}

const CustomTextInput: FC<Props> = ({
  label,
  textInputProps,
  error = "",
  ...props
}) => {
  const [isFocused, setIsFocused] = useState(false);

  const ref = useRef<TextInput>(null);

  return (
    <View {...props}>
      <Text onPress={() => ref?.current?.focus()} style={[styles().label]}>
        {label}
      </Text>
      <View>
        {isFocused && <View style={styles().focusedContainer} />}
        <TextInput
          ref={ref}
          autoCapitalize="none"
          autoCorrect={false}
          style={styles(isFocused).input}
          selectionColor={theme.palette.veryDarkGrey}
          onFocus={() => setIsFocused(true)}
          {...textInputProps}
          onBlur={(e) => {
            setIsFocused(false);
            textInputProps?.onBlur?.(e);
          }}
        />
      </View>
      {!!error && <Text style={styles().error}>{error}</Text>}
    </View>
  );
};

const styles = (isFocused?: boolean) =>
  StyleSheet.create({
    focusedContainer: {
      ...StyleSheet.absoluteFillObject,
      backgroundColor: theme.palette.darkBlue,
      margin: -3,
      borderRadius: theme.borderRadius,
    },
    label: {
      ...theme.typography.regularText,
      color: theme.palette.veryDarkGrey,
      marginBottom: 8,
      alignSelf: "flex-start",
    },
    input: {
      ...theme.typography.regularText,
      lineHeight: 17,
      color: theme.palette.veryDarkGrey,
      borderRadius: theme.borderRadius,
      borderColor: isFocused ? theme.palette.blue : theme.palette.neutral,
      borderWidth: 1,
      paddingHorizontal: 12,
      paddingVertical: 7,
      backgroundColor: theme.palette.white,
    },
    error: {
      ...theme.typography.smallText,
      color: theme.palette.red,
    },
  });

export default CustomTextInput;
