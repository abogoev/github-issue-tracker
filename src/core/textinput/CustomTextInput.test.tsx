import React from "react";
import CustomTextInput from "./CustomTextInput";
import { create, act } from "react-test-renderer";
import { TextInput } from "react-native";

describe("CustomTextInput", () => {
  const label = "Owner";

  it("should render correctly", () => {
    const component = create(<CustomTextInput label={label} />);

    expect(component.toJSON()).toMatchSnapshot("without focus");

    act(() => {
      component.root.findByType(TextInput).props.onFocus();
    });

    expect(component.toJSON()).toMatchSnapshot("with focus");
  });

  it("should write input correctly", () => {
    const onChangeTextMock = jest.fn();
    const component = create(
      <CustomTextInput
        label={label}
        textInputProps={{ onChangeText: onChangeTextMock }}
      />
    );

    const input = "asd";
    component.root.findByType(TextInput).props.onChangeText(input);
    expect(onChangeTextMock).toBeCalledWith(input);
  });

  it("should call onBlur correctly", () => {
    const onBlurMock = jest.fn();
    const component = create(
      <CustomTextInput label={label} textInputProps={{ onBlur: onBlurMock }} />
    );

    expect(onBlurMock).not.toHaveBeenCalled();
    component.root.findByType(TextInput).props.onBlur();
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });
});
