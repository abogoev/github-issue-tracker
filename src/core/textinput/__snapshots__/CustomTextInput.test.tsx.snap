// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextInput should render correctly: with focus 1`] = `
<View>
  <Text
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "color": "#24292f",
          "fontFamily": "Helvetica",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 21,
          "marginBottom": 8,
        },
      ]
    }
  >
    Owner
  </Text>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "rgba(9,105,218,0.3)",
          "borderRadius": 6,
          "bottom": 0,
          "left": 0,
          "margin": -3,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onFocus={[Function]}
      rejectResponderTermination={true}
      selectionColor="#24292f"
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "rgb(9, 105, 218)",
          "borderRadius": 6,
          "borderWidth": 1,
          "color": "#24292f",
          "fontFamily": "Helvetica",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 21,
          "paddingHorizontal": 12,
          "paddingVertical": 5,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      Object {
        "color": "rgb(207, 34, 46)",
        "fontFamily": "Helvetica",
        "fontSize": 12,
        "lineHeight": 18,
      }
    }
  >
    
  </Text>
</View>
`;

exports[`CustomTextInput should render correctly: without focus 1`] = `
<View>
  <Text
    onPress={[Function]}
    style={
      Array [
        Object {
          "alignSelf": "flex-start",
          "color": "#24292f",
          "fontFamily": "Helvetica",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 21,
          "marginBottom": 8,
        },
      ]
    }
  >
    Owner
  </Text>
  <View>
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      onBlur={[Function]}
      onFocus={[Function]}
      rejectResponderTermination={true}
      selectionColor="#24292f"
      style={
        Object {
          "backgroundColor": "#fff",
          "borderColor": "rgb(208, 215, 222)",
          "borderRadius": 6,
          "borderWidth": 1,
          "color": "#24292f",
          "fontFamily": "Helvetica",
          "fontSize": 14,
          "fontWeight": "400",
          "lineHeight": 21,
          "paddingHorizontal": 12,
          "paddingVertical": 5,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <Text
    style={
      Object {
        "color": "rgb(207, 34, 46)",
        "fontFamily": "Helvetica",
        "fontSize": 12,
        "lineHeight": 18,
      }
    }
  >
    
  </Text>
</View>
`;
